src/Servers/Kestrel/shared/HPackHeaderWriter.cs

```diff
@@ -60,20 +72,38 @@ public static bool BeginEncodeHeaders(DynamicHPackEncoder hpackEncoder, Http2Hea
 
         if (!headersEnumerator.MoveNext())
         {
-            return true;
+            return HeaderWriteResult.Done;
+        }
+
+        var done = EncodeHeadersCore(hpackEncoder, headersEnumerator, buffer.Slice(length), canRequestLargerBuffer: true, out var headersLength);
+        length += headersLength;
+        return done;
+    }
+
+    /// <summary>
+    /// Begin encoding headers in the first HEADERS frame without stepping the iterator.
+    /// </summary>
+    public static HeaderWriteResult RetryBeginEncodeHeaders(DynamicHPackEncoder hpackEncoder, Http2HeadersEnumerator headersEnumerator, Span<byte> buffer, out int length)
```

Comment:
I still don't love having a separate method for this.  Some thoughts:
1. If we're going to keep two versions, can they share an implementation (i.e. with a parameter for whether to advance the iterator)?
2. Could we make `BeginEncodeHeaders` _never_ advance the iterator and make that the caller's responsibility?  I guess that would mess up the return value?
3. Unless I'm missing something, `RetryBeginEncodeHeaders` appears to be `BeginEncodeHeaders` without the `MoveNext` - does it also need to skip the `MoveNext` calls in `EncodeHeadersCore`?
4. I know you've explained this, but can you please remind me one more time why we can't just reset the enumerator?  I think it was because an "unknown" enumerator didn't support resetting, but I don't think I fully grasped what that enumerator was and why it doesn't (and can't) support that.

Reply:
Was there a reply to the non-perf part of https://github.com/dotnet/aspnetcore/pull/55322#issuecomment-2136328936?  I can't seem to find it.

Reply:
I think 1. is doable, let me update the PR.

The original reason for creating this was: trailers can return (with a single large trailer) "BufferToSmall" response. At this point, I could not 'Reset' the state of the headersenumerator when it also contained the values in the 'unknown collection': https://github.com/dotnet/aspnetcore/pull/55322#issuecomment-2136680427

The other options:

2. seems odd in the sense that it is `HPackHeaderWriter` only who calls `MoveNext()` on the headers enumerator, so I would find it weird to share this responsibility and to expect the caller to do this.

3. No because `EncodeHeadersCore` returns before it would call `MoveNext()` (it has this do-while loop), so in the case it would return `BufferTooSmall`, it does not move the iterator.

4. https://github.com/dotnet/aspnetcore/pull/55322#issuecomment-2136680427 I don't think the `Reset` is used anywhere else, I learned not long ago that many `IEnumerable<T>` implementations simply throw on `Reset()` in the BCL. I don't see a reason against HeadersEnumerator resetting the "unknown" collection, but I also wasn't sure breaking anything if that is added.